Fix FormTagHelper to set flag indicating generation of AntiforgeryToken

kichalla opened the pull request:
[Fixes #4595] Better error message with extraneous @Html.AntiforgeryToken

@Eilon @dougbu

kichalla commented on test/Microsoft.AspNetCore.Mvc.FunctionalTests/TagHelpersTest.cs:
```diff
@@ -87,6 +87,42 @@ public class TagHelpersTest :
+        public async Task ReregisteringAntiforgeryTokenInsideFormTagHelper_DoesNotAddDuplicateAntifogeryTokenFields()
+        {
+            // Arrange
+            var expectedMediaType = MediaTypeHeaderValue.Parse("text/html; charset=utf-8");
```
Mostly copy pasted from a `HtmlGenerationTest`

kichalla commented on test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/FormTagHelperTest.cs:
```diff
@@ -92,6 +92,7 @@ public class FormTagHelperTest
             Assert.Equal("post", attribute.Value);
             attribute = Assert.Single(output.Attributes, attr => attr.Name.Equals("action"));
             Assert.Equal("home/index", attribute.Value);
+            Assert.True(viewContext.FormContext.HasAntiforgeryToken);
```
This check failed before my fix

kichalla commented:
oops..meant to add @rynowak and @dougbu :smile:

rynowak commented on test/Microsoft.AspNetCore.Mvc.FunctionalTests/TagHelpersTest.cs:
```diff
@@ -87,6 +87,42 @@ public class TagHelpersTest :
         }
 
+        [Fact]
+        public async Task ReregisteringAntiforgeryTokenInsideFormTagHelper_DoesNotAddDuplicateAntifogeryTokenFields()
```
If you're going to misspell it, you might as well misspell it right.

![image](https://cloud.githubusercontent.com/assets/1430011/15231348/684c1bfe-184f-11e6-8e65-df5f896fe0c1.png)

kichalla replied:
You got a good eye at catching the antiforgery spelling :-)

dougbu replied:
Which one is John?

rynowak commented on test/Microsoft.AspNetCore.Mvc.FunctionalTests/TagHelpersTest.cs:
```diff
@@ -87,6 +87,42 @@ public class TagHelpersTest :
+                await ResourceFile.ReadResourceAsync(_resourcesAssembly, outputFile, sourceFile: false);
+
+            // Act
+            // The host is not important as everything runs in memory and tests are isolated from each other.
```
why is this comment here?

rynowak commented on src/Microsoft.AspNetCore.Mvc.TagHelpers/FormTagHelper.cs:
```diff
@@ -240,6 +240,7 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
                 var antiforgeryTag = Generator.GenerateAntiforgery(ViewContext);
                 if (antiforgeryTag != null)
                 {
+                    ViewContext.FormContext.HasAntiforgeryToken = true;
```
Can we add a unit test for this?

kichalla replied:
Not sure if you noticed but I updated an existing test with a condition https://github.com/aspnet/Mvc/pull/4626/files/39f290193310e15bc3cb050092de85da4dc378e5#diff-c3f8a68da4023116edc455c5e781fc80R95

Would it suffice?

rynowak replied:
Oh! yeah you're right, you did add that. Ok

rynowak replied:
This should really have it's own test **OR** be a part of all of them.

kichalla replied:
I chose to update all of them as they seem to cover different scenarios.

rynowak commented on test/WebSites/TagHelpersWebSite/Views/Employee/DuplicateAntiforgeryTokenRegistration.cshtml:
```diff
@@ -0,0 +1,11 @@
+}
+<html>
+<body>
+    <!--Form tag helper already registers antiforgery token. The Html helper call should be no-op-->
```
If you use a razor comment it won't go to the output

rynowak commented:
:watch:

kichalla commented:
:up: :date:

rynowak commented:
:shipit:

dougbu commented on src/Microsoft.AspNetCore.Mvc.TagHelpers/FormTagHelper.cs:
```diff
@@ -240,6 +240,7 @@ public override void Process(TagHelperContext context, TagHelperOutput output)
                 var antiforgeryTag = Generator.GenerateAntiforgery(ViewContext);
```
Why not update the `HasAntiforgeryToken` property in the `GenerateAntiforgery()` method? No longer have any cases where it's _not_ set.

dougbu commented:
Like the tests!!

dougbu commented:
:watch: at least for setting `HasAntiforgeryToken` once. As discussed offline, let's see if it works to go further and _read_ the property just one place.

kichalla commented:
:up: :date:

kichalla commented on test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Rendering/HtmlHelperFormTest.cs:
```diff
@@ -434,50 +433,6 @@ public void BeginForm_EndForm_RendersAntiforgeryTokenWhenMethodIsPost()
-            {
-                Assert.True(viewContext.FormContext.HasAntiforgeryToken);
-
-                // This call will no-op
```
Since this is no longer a criteria that needs to be tested from a HtmlHelper perspective and also since its being covered in DefaultHtmlGenerator, we can delete this test.

dougbu commented on test/WebSites/TagHelpersWebSite/Views/Employee/DuplicateAntiforgeryTokenRegistration.cshtml:
```diff
@@ -0,0 +1,11 @@
+<html>
+<body>
+    @*Form tag helper already registers antiforgery token. The Html helper call should be no-op*@
+    <form asp-antiforgery="true" asp-action="DuplicateAntiforgeryTokenRegistration">@Html.AntiForgeryToken()</form>
```
What is this test for? Should be covered completely by the first `<form>` in the other new `.cshtml` file. I'm also very surprised we don't already have `Html.AntiForgeryToken()` calls in our functional test sites.

kichalla replied:
This was the actual bug that @sebastienros  ran into (in his case the `asp-antiforgery` attribute was _not_ used...I added it explicitly so as not depend on some implicit behavior)

dougbu replied:
:ok:

dougbu commented on test/WebSites/HtmlGenerationWebSite/Views/HtmlGeneration_Home/DuplicateAntiforgeryTokenRegistration.cshtml:
```diff
@@ -0,0 +1,19 @@
+﻿@{
```
Rename everything about this new test to reflect the actual scenario. Not testing any duplication, just multiple `<form>`s in a single view.

kichalla replied:
> just multiple <form>s in a single view.

I believe we already have tests for this scenario. This is specifically for scenario where user is trying to call `@Html.AntiForgeryToken()` even though the beginform/beginrouteform already do that. I did not want to create 2 separate files for this negative scenario and hence clubbed them into one.

dougbu replied:
Bet we didn't remove `Html.AntiForgeryToken()` calls when we changed the `BeginForm()` / `BeginRouteForm()` default. So that part is likely covered. This leaves the multiple `<form>`s (unlikely covered elsewhere) and use of `BeginRouteForm()` as the potentially-unique bits.

dougbu commented on test/Microsoft.AspNetCore.Mvc.FunctionalTests/HtmlGenerationTest.cs:
```diff
@@ -530,6 +530,43 @@ public class HtmlGenerationTest :
+                // Reverse usual substitution and insert a format item into the new file content.
+                responseContent = responseContent.Replace(beginFormAntiforgeryToken, "{0}");
+                responseContent = responseContent.Replace(beginRouteFormAntiforgeryToken, "{1}");
+                ResourceFile.UpdateFile(_resourcesAssembly, outputFile, expectedContent, responseContent);
```
De-indent to make this consistent w/ rest of the method.

dougbu commented on test/Microsoft.AspNetCore.Mvc.FunctionalTests/HtmlGenerationTest.cs:
```diff
@@ -530,6 +530,43 @@ public class HtmlGenerationTest :
         }
 
+        [Fact]
+        public async Task ReregisteringAntiforgeryInBeginFormAndBeginRouteForm_DoesNotAddDuplicateAntiforgeryTokens()
```
Please simplify this test name e.g. `MultipleFormElements_DoNotConfuseTheSystem` :trollface: or `MultipleFormsAndAntiForgeryTokens_AddOneTokenPerForm()`.

Eilon replied:
I'm pretty sure this name is just fine the way it is...

dougbu commented on test/Microsoft.AspNetCore.Mvc.FunctionalTests/TagHelpersTest.cs:
```diff
@@ -87,6 +87,41 @@ public class TagHelpersTest :
+#if GENERATE_BASELINES
+                // Reverse usual substitution and insert a format item into the new file content.
+                responseContent = responseContent.Replace(forgeryToken, "{0}");
+                ResourceFile.UpdateFile(_resourcesAssembly, outputFile, expectedContent, responseContent);
```
Fix indentation here too.

dougbu commented:
:watch: for some test cleanup

kichalla commented:
:up: :date: I deleted the test we talked about as I found @harshgMSFT's test that he added for multiple forms in a single page. I added a comment in that file.

kichalla commented:
@dougbu ping

dougbu commented:
:shipit:
